lru_cache annotation doesn't work as property

So I originally reported it in typeshed (#2659) but was told this can't be fixed there and it is mypy limitation so repporting it here.

----

I have following code:

```python
#! /usr/bin/env python

import time
from functools import lru_cache

class MyClass:
	@property
	@lru_cache(1)
	def expensive_operation(self) -> int:
		time.sleep(2)
		return 42

	def another_operation(self) -> int:
		return self.expensive_operation + 1

if __name__ == '__main__':
	o = MyClass()
	print(o.expensive_operation)
	print(o.expensive_operation)
	print(o.another_operation())
```

It works as expected:
```
$ ./test.py
42
42
43
```

Yet, mypy is showing errors:
```
$ mypy test.py
test.py:7: error: Decorated property not supported
test.py:14: error: Unsupported operand types for + ("_lru_cache_wrapper[int]" and "int")
```

The `lru_cache` works just fine when decorated as property, and the returned type should be `int` and not `_lru_cache_wrapper[int]`.

This is on python 3.7.0 and mypy 0.641.

@probablyodd well its a bit different. If you set `# type: ignore` to `@property` in this case, 
mypy will completely ignore existence of this decorator and treat such methods as common callable methods. Mypy will just say somethingh like `"lru_cache" has no attribute "..."` if complex object such as dict returned from property method.

```
from functools import lru_cache

class Foo:
    @property  # type: ignore
    @lru_cache
    def bar(self) -> dict:
        return dict(key='value')

Foo().bar.get('key')
```

```
test.py:11: error: "lru_cache" has no attribute "get"
```

If there is no plan to fix this soon, then can this be documented properly? This error is implicated by [the recommendation in the standard library docs for `functools.cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) for caching properties in classes without proper `__dict__`s, but there's no hint that such recommendation would frustrate Mypy:

``` python
# test_5858.py
import statistics
from functools import cache

class DataSet:
    __slots__ = ("_data",)

    def __init__(self, sequence_of_numbers):
        self._data = sequence_of_numbers

    @property  # <- error: Decorated property not supported  [misc]
    @cache  # <- results in downstream errors involving type "_lru_cache_wrapper[…]"
    def stdev(self):
        return statistics.stdev(self._data)

f: float = statistics.stdev(range(10))
f = DataSet(range(10)).stdev
```

``` sh
% python --version
Python 3.9.10
% mypy --version
mypy 0.941
% mypy --config-file=/dev/null test_5858.py
/dev/null: No [mypy] section in config file
test_5858.py:11: error: Decorated property not supported
test_5858.py:17: error: Incompatible types in assignment (expression has type "_lru_cache_wrapper[Any]", variable has type "float")
Found 2 errors in 1 file (checked 1 source file)
```

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -780,7 +780,13 @@ def analyze_var(
             var.is_initialized_in_class
             and (not is_instance_var(var) or mx.is_operator)
         ):
-            call_type = _analyze_member_access("__call__", typ, mx)
+            call_type: ProperType
+            if isinstance(typ, FunctionLike) and not typ.is_type_obj():
+                call_type = typ
+            elif var.is_property:
+                call_type = get_proper_type(_analyze_member_access("__call__", typ, mx))
+            else:
+                call_type = typ
             if isinstance(call_type, FunctionLike) and not call_type.is_type_obj():
                 if mx.is_lvalue:
                     if var.is_property:

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -2,7 +2,7 @@
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Callable, Sequence, cast
+from typing import TYPE_CHECKING, Callable, Optional, Sequence, cast
 
 from mypy import meet, message_registry, subtypes
 from mypy.erasetype import erase_typevars
@@ -776,50 +776,52 @@ def analyze_var(
         freeze_all_type_vars(t)
         result: Type = t
         typ = get_proper_type(typ)
+
+        call_type: Optional[ProperType] = None
         if var.is_initialized_in_class and (not is_instance_var(var) or mx.is_operator):
-            call_type: ProperType
             if isinstance(typ, FunctionLike) and not typ.is_type_obj():
                 call_type = typ
             elif var.is_property:
                 call_type = get_proper_type(_analyze_member_access("__call__", typ, mx))
             else:
                 call_type = typ
-            if isinstance(call_type, FunctionLike) and not call_type.is_type_obj():
-                if mx.is_lvalue:
-                    if var.is_property:
-                        if not var.is_settable_property:
-                            mx.msg.read_only_property(name, itype.type, mx.context)
-                    else:
-                        mx.msg.cant_assign_to_method(mx.context)
 
-                if not var.is_staticmethod:
-                    # Class-level function objects and classmethods become bound methods:
-                    # the former to the instance, the latter to the class.
-                    functype = call_type
-                    # Use meet to narrow original_type to the dispatched type.
-                    # For example, assume
-                    # * A.f: Callable[[A1], None] where A1 <: A (maybe A1 == A)
-                    # * B.f: Callable[[B1], None] where B1 <: B (maybe B1 == B)
-                    # * x: Union[A1, B1]
-                    # In `x.f`, when checking `x` against A1 we assume x is compatible with A
-                    # and similarly for B1 when checking against B
-                    dispatched_type = meet.meet_types(mx.original_type, itype)
-                    signature = freshen_all_functions_type_vars(functype)
-                    bound = get_proper_type(expand_self_type(var, signature, mx.original_type))
-                    assert isinstance(bound, FunctionLike)
-                    signature = bound
-                    signature = check_self_arg(
-                        signature, dispatched_type, var.is_classmethod, mx.context, name, mx.msg
-                    )
-                    signature = bind_self(signature, mx.self_type, var.is_classmethod)
-                    expanded_signature = expand_type_by_instance(signature, itype)
-                    freeze_all_type_vars(expanded_signature)
-                    if var.is_property:
-                        # A property cannot have an overloaded type => the cast is fine.
-                        assert isinstance(expanded_signature, CallableType)
-                        result = expanded_signature.ret_type
-                    else:
-                        result = expanded_signature
+        if isinstance(call_type, FunctionLike) and not call_type.is_type_obj():
+            if mx.is_lvalue:
+                if var.is_property:
+                    if not var.is_settable_property:
+                        mx.msg.read_only_property(name, itype.type, mx.context)
+                else:
+                    mx.msg.cant_assign_to_method(mx.context)
+
+            if not var.is_staticmethod:
+                # Class-level function objects and classmethods become bound methods:
+                # the former to the instance, the latter to the class.
+                functype: FunctionLike = call_type
+                # Use meet to narrow original_type to the dispatched type.
+                # For example, assume
+                # * A.f: Callable[[A1], None] where A1 <: A (maybe A1 == A)
+                # * B.f: Callable[[B1], None] where B1 <: B (maybe B1 == B)
+                # * x: Union[A1, B1]
+                # In `x.f`, when checking `x` against A1 we assume x is compatible with A
+                # and similarly for B1 when checking against B
+                dispatched_type = meet.meet_types(mx.original_type, itype)
+                signature = freshen_all_functions_type_vars(functype)
+                bound = get_proper_type(expand_self_type(var, signature, mx.original_type))
+                assert isinstance(bound, FunctionLike)
+                signature = bound
+                signature = check_self_arg(
+                    signature, dispatched_type, var.is_classmethod, mx.context, name, mx.msg
+                )
+                signature = bind_self(signature, mx.self_type, var.is_classmethod)
+                expanded_signature = expand_type_by_instance(signature, itype)
+                freeze_all_type_vars(expanded_signature)
+                if var.is_property:
+                    # A property cannot have an overloaded type => the cast is fine.
+                    assert isinstance(expanded_signature, CallableType)
+                    result = expanded_signature.ret_type
+                else:
+                    result = expanded_signature
     else:
         if not var.is_ready and not mx.no_deferral:
             mx.not_ready_callback(var.name, mx.context)

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -779,44 +779,44 @@ def analyze_var(
         if (
             var.is_initialized_in_class
             and (not is_instance_var(var) or mx.is_operator)
-            and isinstance(typ, FunctionLike)
-            and not typ.is_type_obj()
         ):
-            if mx.is_lvalue:
-                if var.is_property:
-                    if not var.is_settable_property:
-                        mx.msg.read_only_property(name, itype.type, mx.context)
-                else:
-                    mx.msg.cant_assign_to_method(mx.context)
+            call_type = _analyze_member_access("__call__", typ, mx)
+            if isinstance(call_type, FunctionLike) and not call_type.is_type_obj():
+                if mx.is_lvalue:
+                    if var.is_property:
+                        if not var.is_settable_property:
+                            mx.msg.read_only_property(name, itype.type, mx.context)
+                    else:
+                        mx.msg.cant_assign_to_method(mx.context)
 
-            if not var.is_staticmethod:
-                # Class-level function objects and classmethods become bound methods:
-                # the former to the instance, the latter to the class.
-                functype = typ
-                # Use meet to narrow original_type to the dispatched type.
-                # For example, assume
-                # * A.f: Callable[[A1], None] where A1 <: A (maybe A1 == A)
-                # * B.f: Callable[[B1], None] where B1 <: B (maybe B1 == B)
-                # * x: Union[A1, B1]
-                # In `x.f`, when checking `x` against A1 we assume x is compatible with A
-                # and similarly for B1 when checking against B
-                dispatched_type = meet.meet_types(mx.original_type, itype)
-                signature = freshen_all_functions_type_vars(functype)
-                bound = get_proper_type(expand_self_type(var, signature, mx.original_type))
-                assert isinstance(bound, FunctionLike)
-                signature = bound
-                signature = check_self_arg(
-                    signature, dispatched_type, var.is_classmethod, mx.context, name, mx.msg
-                )
-                signature = bind_self(signature, mx.self_type, var.is_classmethod)
-                expanded_signature = expand_type_by_instance(signature, itype)
-                freeze_all_type_vars(expanded_signature)
-                if var.is_property:
-                    # A property cannot have an overloaded type => the cast is fine.
-                    assert isinstance(expanded_signature, CallableType)
-                    result = expanded_signature.ret_type
-                else:
-                    result = expanded_signature
+                if not var.is_staticmethod:
+                    # Class-level function objects and classmethods become bound methods:
+                    # the former to the instance, the latter to the class.
+                    functype = call_type
+                    # Use meet to narrow original_type to the dispatched type.
+                    # For example, assume
+                    # * A.f: Callable[[A1], None] where A1 <: A (maybe A1 == A)
+                    # * B.f: Callable[[B1], None] where B1 <: B (maybe B1 == B)
+                    # * x: Union[A1, B1]
+                    # In `x.f`, when checking `x` against A1 we assume x is compatible with A
+                    # and similarly for B1 when checking against B
+                    dispatched_type = meet.meet_types(mx.original_type, itype)
+                    signature = freshen_all_functions_type_vars(functype)
+                    bound = get_proper_type(expand_self_type(var, signature, mx.original_type))
+                    assert isinstance(bound, FunctionLike)
+                    signature = bound
+                    signature = check_self_arg(
+                        signature, dispatched_type, var.is_classmethod, mx.context, name, mx.msg
+                    )
+                    signature = bind_self(signature, mx.self_type, var.is_classmethod)
+                    expanded_signature = expand_type_by_instance(signature, itype)
+                    freeze_all_type_vars(expanded_signature)
+                    if var.is_property:
+                        # A property cannot have an overloaded type => the cast is fine.
+                        assert isinstance(expanded_signature, CallableType)
+                        result = expanded_signature.ret_type
+                    else:
+                        result = expanded_signature
     else:
         if not var.is_ready and not mx.no_deferral:
             mx.not_ready_callback(var.name, mx.context)

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -776,10 +776,7 @@ def analyze_var(
         freeze_all_type_vars(t)
         result: Type = t
         typ = get_proper_type(typ)
-        if (
-            var.is_initialized_in_class
-            and (not is_instance_var(var) or mx.is_operator)
-        ):
+        if var.is_initialized_in_class and (not is_instance_var(var) or mx.is_operator):
             call_type: ProperType
             if isinstance(typ, FunctionLike) and not typ.is_type_obj():
                 call_type = typ

--- a/test-data/unit/check-functions.test
+++ b/test-data/unit/check-functions.test
@@ -3158,3 +3158,20 @@ class C(A, B):
 class D(A, B):
     def f(self, z: int) -> str: pass  # E: Method "f" is not using @override but is overriding a method in class "__main__.A"
 [typing fixtures/typing-override.pyi]
+
+[case testCallableProperty]
+from typing import Callable
+
+class something_callable:
+    def __call__(self, fn) -> str: ...
+
+def decorator(fn: Callable[..., int]) -> something_callable: ...
+
+class A:
+    @property
+    @decorator
+    def f(self) -> int: ...
+
+reveal_type(A.f)  # N: Revealed type is "__main__.something_callable"
+reveal_type(A().f)  # N: Revealed type is "builtins.str"
+[builtins fixtures/property.pyi]

